src/libstd/sys/unix/stack_overflow.rs

```diff
@@ -169,13 +169,27 @@ mod imp {
     }
 
     pub unsafe fn make_handler() -> Handler {
-        let stack = get_stack();
-        sigaltstack(&stack, ptr::null_mut());
-        Handler { _data: stack.ss_sp as *mut libc::c_void }
+        let mut stack = mem::zeroed();
+        sigaltstack(ptr::null(), &mut stack);
+        if stack.ss_flags & SS_DISABLE != 0 {
+            stack = get_stack();
+            sigaltstack(&stack, ptr::null_mut());
+            Handler { _data: stack.ss_sp as *mut libc::c_void }
+        } else {
+            Handler { _data: ptr::null_mut() }
+        }
     }
 
     pub unsafe fn drop_handler(handler: &mut Handler) {
-        munmap(handler._data, SIGSTKSZ);
+        if !handler._data.is_null() {
+            let stack =  libc::stack_t {
+                ss_sp: ptr::null_mut(),
+                ss_flags: SS_DISABLE,
+                ss_size: 0
+            };
+            sigaltstack(&stack, ptr::null_mut());
```

Comment:
The alternate stack is configured per thread, so only thread that could use
this stack is the current one. In other words, the stack would be in use if and
only if this code would be executed from a signal handler.

Generally any errors returned from sigaltsack would be completely unexpected.
And I didn't even bother to check them for this reason. Though, if I were to
add checks, then only response to them would be to abort the program, because
it means that we have committed some serious programming error.